### PR TITLE
Python wrapper classes for NASA9 and piecewise Gibbs

### LINKED ISSUE #705: Python SpeciesThermo wrapper classes missing for NASA9 and piecewise Gibbs (Mu0Poly)
In `speciesthermo.pyx`, wrapper classes are defined for the most commonly used species thermo models (`NasaPoly2`, `ShomatePoly2`, and `ConstantCp`) but not for the other available models (`Nasa9PolyMultiTempRegion` and `Mu0Poly`).

### PR DESCRIPTION
**Pull request by ischoegl**
Please fill in the issue number this pull request is fixing:

Fixes #705

Changes proposed in this pull request:
- Add `Nasa9PolyMultiTempRegion` and `Mu0Poly` class wrappers to `speciesthermo.pyx`
- Add missing C++ constructor for `Nasa9PolyMultiTempRegion`
- Enable variable number of coefficients
- Add unit tests

Note: PR depends on #720 (`piecewise-Gibbs` is affected by #595)

### CODE REVIEW

**Comment by ischoegl** [comment]
@speth ... I wasn’t able to find a reference for a `Mu0Poly` YAML entry ... i am planning to add a test to this PR (adding the entry to `thermo-models.yaml`). I found the `AnyMap` parser, 
https://github.com/Cantera/cantera/blob/55fe2233b18d72aa60d776d2bdace14f98d94346/src/thermo/SpeciesThermoFactory.cpp#L368
but would like to avoid guessing the correct format.

**Comment by ischoegl** [comment]
Added another test after locating a `piecewise-Gibbs` YAML entry, which fails with:
```
======================================================================
FAIL: test_piecewise_Gibbs_load (cantera.test.test_thermo.TestSpeciesThermo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/src/build/python/cantera/test/test_thermo.py", line 1204, in test_piecewise_Gibbs_load
    sol = ct.Solution('thermo-models.yaml', 'HMW-NaCl-electrolyte')
  File "interfaces/cython/cantera/thermo.pyx", line 267, in cantera._cantera.ThermoPhase.__init__
  File "interfaces/cython/cantera/transport.pyx", line 140, in cantera._cantera.Transport.__init__
cantera._cantera.CanteraError: 
***********************************************************************
CanteraError thrown by HMWSoln::setDensity:
Density is not an independent variable
***********************************************************************
```
indicating that there is a co-dependence with #720 and #595. Specifically, there is a call to `Phase::restoreState` here:
https://github.com/Cantera/cantera/blob/55fe2233b18d72aa60d776d2bdace14f98d94346/src/transport/TransportFactory.cpp#L91

Update: issue is fixed in #720

**Comment by bryanwweber** [comment]
Interestingly, I ran into exactly this same error today in #693. Can you extract the fix into its own PR so it can be merged a little quicker?

**Comment by ischoegl** [comment]
> Interestingly, I ran into exactly this same error today in #693. Can you extract the fix into its own PR so it can be merged a little quicker?

@bryanwweber that's indeed curious! ... I worked on #720 for a while today (streamlining & simplifying, so it covers all of `thermo-models.yaml`): it's close, and should fix all of the behavior covered in #595 for good (I expect that you'll find this pop up multiple times in #693). Only thing missing is `lattice-compound` (which disables setting of mass fractions). I.e. once this is working correctly, the entire PR should be reviewed.

Edit: #720 is ready for review

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (speth) [inline interfaces/cython/cantera/speciesthermo.pyx]: Aren't the coefficients nondimensional, since the polynomials are for c_p/R and H/RT, as per https://cantera.org/science/science-species.html#the-nasa-9-coefficient-polynomial-parameterization?

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (speth) [inline interfaces/cython/cantera/speciesthermo.pyx]: This makes it seem like `coeffs` can be a nested array. I think this needs to be written out in more gorey detail, i.e.
`coeffs[1 + 11*zone]: minimum temperature` (or whatever the correct index is).

File: include/cantera/thermo/Nasa9PolyMultiTempRegion.h
Comment (speth) [inline include/cantera/thermo/Nasa9PolyMultiTempRegion.h]: The structure of `coeffs` needs to be documented here.

File: include/cantera/thermo/Nasa9PolyMultiTempRegion.h
Comment (speth) [inline include/cantera/thermo/Nasa9PolyMultiTempRegion.h]: Indentation error, I think?

File: include/cantera/thermo/SpeciesThermoInterpType.h
Comment (speth) [inline include/cantera/thermo/SpeciesThermoInterpType.h]:
```suggestion
    //! This utility function returns the number of coefficients
```

File: src/thermo/Nasa9PolyMultiTempRegion.cpp
Comment (speth) [inline src/thermo/Nasa9PolyMultiTempRegion.cpp]: We usually try (admittedly not consistent, especially in older code) to use `static_cast` rather than the C-style casts, i.e.
```suggestion
    size_t regions = static_cast<size_t>(coeffs[0]);
```

File: src/thermo/Nasa9PolyMultiTempRegion.cpp
Comment (speth) [inline src/thermo/Nasa9PolyMultiTempRegion.cpp]: I think this can be simplified a bit by passing arguments into the constructor to `Nasa9Poly1`, e.g.
```c++
Nasa9Poly1* poly = new Nasa9Poly1(coeffs[11*i+1], coeffs[11*i+2], pref, coeffs + 11*i + 3);
```

File: src/thermo/Nasa9PolyMultiTempRegion.cpp
Comment (speth) [inline src/thermo/Nasa9PolyMultiTempRegion.cpp]: I would recommend delegating initialization of these three properties through a call to the `SpeciesThermoInterpType` constructor. For example, see the equivalent constructor for `NasaPoly2`.

File: src/thermo/SpeciesThermoInterpType.cpp
Comment (speth) [inline src/thermo/SpeciesThermoInterpType.cpp]: You can use `NotImplementedError("SpeciesThermoInterpType::nCoeffs")` here.

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (speth) [inline interfaces/cython/cantera/speciesthermo.pyx]: A docstring for this function would be useful.

Also, we could avoid the need to override this in every child class by having `return n == self.n_coeffs` be the default implementation, which would be correct for all classes where the number of coefficients is a constant.

**Review by speth — CHANGES_REQUESTED** [review]
Thanks, this clears up an obvious gap in the Python interface. I think all of my comments are just about implementation details and documentation.

**Comment by ischoegl** [comment]
@speth: thank you for the review! I will, however, wait for #720 to make sure that things can be merged in order.

**Comment by speth** [comment]
Well, I think my review of #720 is going to be a little more involved, so I wouldn't mind getting this merged first. The test errors here can be fixed quite easily, by replacing `Solution` with `ThermoPhase` in the two failing tests.

File: src/thermo/SpeciesThermoInterpType.cpp
Comment (ischoegl) [reply src/thermo/SpeciesThermoInterpType.cpp]: True. Had copied this from neighboring functions. Updating all 5 instances.

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (ischoegl) [reply interfaces/cython/cantera/speciesthermo.pyx]: Correct. Looks like I copied a non-pertinent docstring fragment.

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (ischoegl) [reply interfaces/cython/cantera/speciesthermo.pyx]: Ok on the docstring.

The other suggestion won’t work all the time as the number of zones is not known a priori for the NASA9 multi temp case.

File: src/thermo/Nasa9PolyMultiTempRegion.cpp
Comment (ischoegl) [reply src/thermo/Nasa9PolyMultiTempRegion.cpp]: indeed.

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (ischoegl) [reply interfaces/cython/cantera/speciesthermo.pyx]: Ok to docstring.

The suggestion won't work as the function is called prior to the instantiation of the C++ object. This is clarified in the docstring.

**Comment by ischoegl** [comment]
NP. Adopted suggested changes and (temporarily) replaced `Solution` by `ThermoPhase` in the two failing tests. Can switch this back after rebasing #720 upon merge of this PR.

File: interfaces/cython/cantera/speciesthermo.pyx
Comment (speth) [reply interfaces/cython/cantera/speciesthermo.pyx]: Ah, fair point. The documentation of the tricky behavior is appreciated.

**Review by speth — APPROVED** [review]

**Pull request merged**
